Fix bad parameter count in code generator

dain opened the pull request:
Avoid BLOCK_POSITION when there a lot of function arguments

findepi reviewed (commented):
Please add a unit test.
Also, please add an integration test ensuring there isn't a similar limitation elsewhere.

wendigo commented:
One of the tests is failing right now:

```
2021-02-24T04:48:55.6686772Z [ERROR] testVersionOnCompilerFailedError(io.trino.tests.TestServer)  Time elapsed: 0.878 s  <<< FAILURE!
2021-02-24T04:48:55.6690974Z java.lang.RuntimeException: Error expected
2021-02-24T04:48:55.6696818Z 	at io.trino.tests.TestServer.lambda$checkVersionOnError$16(TestServer.java:289)
2021-02-24T04:48:55.6701969Z 	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
2021-02-24T04:48:55.6707765Z 	at io.trino.tests.TestServer.checkVersionOnError(TestServer.java:289)
2021-02-24T04:48:55.6714478Z 	at io.trino.tests.TestServer.testVersionOnCompilerFailedError(TestServer.java:278)
2021-02-24T04:48:55.6721289Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2021-02-24T04:48:55.6729674Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2021-02-24T04:48:55.6738202Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2021-02-24T04:48:55.6748072Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2021-02-24T04:48:55.6759858Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2021-02-24T04:48:55.6767042Z 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
2021-02-24T04:48:55.6775082Z 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
2021-02-24T04:48:55.6781653Z 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
2021-02-24T04:48:55.6789872Z 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
2021-02-24T04:48:55.6796166Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
2021-02-24T04:48:55.6803433Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2021-02-24T04:48:55.6810765Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2021-02-24T04:48:55.6815305Z 	at java.base/java.lang.Thread.run(Thread.java:834)
```

dain commented:
@findepi I fixed the test failure... it was forcing a compiler failure using the same pivot technique.  I changed it to another query that also fails during compilation.  @wendigo has a PR with an additional test for this case.

findepi commented on testing/trino-tests/src/test/java/io/trino/tests/TestServer.java:
```diff
@@ -257,25 +258,21 @@ public void testVersionOnError()
-            checkVersionOnError(pivotQuery, "TrinoException: Compiler failed(?s:.*)at io.trino.sql.gen.PageFunctionCompiler.compileProjection");
+            testingClient.execute("CREATE TABLE " + tableName + " AS SELECT '' as foo");
+
+            // This query is designed to cause a compile failure, and hopefully not run to long.
```
to -> too

findepi commented on testing/trino-tests/src/test/java/io/trino/tests/TestServer.java:
```diff
@@ -257,25 +258,21 @@ public void testVersionOnError()
+            testingClient.execute("CREATE TABLE " + tableName + " AS SELECT '' as foo");
+
+            // This query is designed to cause a compile failure, and hopefully not run to long.
+            // The exact query is not important, just that it causes a failure.
```
Seems redundant, this is what test method name is saying

dain replied:
I like comments like this as it helps me when updating to know if the exact query shape was important.

findepi commented on testing/trino-tests/src/test/java/io/trino/tests/TestServer.java:
```diff
@@ -257,25 +258,21 @@ public void testVersionOnError()
+            String array = IntStream.range(0, 254)
+                    .mapToObj(columnNumber -> "foo")
+                    .collect(joining(", ", "ARRAY[", "]"));
+            String pivotQuery = "SELECT " +
```
pivotQuery -> query

findepi commented on testing/trino-tests/src/test/java/io/trino/tests/TestServer.java:
```diff
@@ -257,25 +258,21 @@ public void testVersionOnError()
-                            .mapToObj(columnNumber -> format("f%s", columnNumber))
-                            .collect(joining(", ", "ARRAY[", "]")) + ") t(x, y)";
-            checkVersionOnError(pivotQuery, "TrinoException: Compiler failed(?s:.*)at io.trino.sql.gen.PageFunctionCompiler.compileProjection");
+            testingClient.execute("CREATE TABLE " + tableName + " AS SELECT '' as foo");
```
I guess the query can select from VALUES and thus you don't need to DROP and CREATE here

dain replied:
Using values is dangerous as it can be optimized away.  I'm just going to leave the query as is.
